src/uri/uri.cc

```diff
@@ -70,25 +73,83 @@ struct URI::Internal {
 };
 
 URI::URI(std::string input) : data{std::move(input)}, internal{new Internal} {
-  uri_parse(this->data, &this->internal->uri);
+  this->parse();
 }
 
 URI::URI(std::istream &input) : internal{new Internal} {
   std::ostringstream output;
   output << input.rdbuf();
   this->data = output.str();
-  uri_parse(this->data, &this->internal->uri);
+  this->parse();
 }
 
 URI::~URI() { uriFreeUriMembersA(&this->internal->uri); }
 
+// TODO: Test the copy constructor
 URI::URI(const URI &other) : URI{other.recompose()} {}
 
 URI::URI(URI &&other)
     : data{std::move(other.data)}, internal{std::move(other.internal)} {
+  this->parsed = other.parsed;
+  this->path_components_ = std::move(other.path_components_);
+  this->scheme_ = std::move(other.scheme_);
+  this->userinfo_ = std::move(other.userinfo_);
+  this->host_ = std::move(other.host_);
+  this->port_ = std::move(other.port_);
+  this->fragment_ = std::move(other.fragment_);
+  this->query_ = std::move(other.query_);
```

Comment:
I wonder if we should not have a URIElement struct to abstract of all these.

Reply:
I think it's fine as it is. Very simple and direct